File: cli/src/config/loadRouting.ts
import { parse as parseYaml } from 'yaml';
import type { FetchMethod } from '../types/capture.js';

export type StrategyName = FetchMethod;

export type ApifyRouteConfig = {
  actorId: string;
  build?: string;
  inputFromUrl?: boolean;
  /**
   * YouTube-only: Apify actors differ on input shape.
   * - `start_urls` — `{ startUrls: [{ url }] }` (website-style crawlers)
   * - `urls` — `{ urls: [url] }` (e.g. `automation-lab/youtube-transcript` on Apify Store)
   */
  youtubeInput?: 'start_urls' | 'urls';
};

export type RouteMatch = {
  hostSuffix?: string;
  pathPrefix?: string;
};

export type RoutingRoute = {
  match: RouteMatch;
  strategy: StrategyName;
  apify?: ApifyRouteConfig;
};

export type RoutingConfig = {
  version: number;
  defaultStrategy: StrategyName;
  routes: RoutingRoute[];
  apifyDefaults?: ApifyRouteConfig;
};

export function loadRouting(yamlText: string): RoutingConfig {
  const raw = parseYaml(yamlText) as RoutingConfig;
  if (raw?.version !== 1) throw new Error('routing: expected version: 1');
  if (!raw.defaultStrategy) throw new Error('routing: missing defaultStrategy');
  if (!Array.isArray(raw.routes)) throw new Error('routing: routes must be an array');
  return raw;
}

function hostMatches(hostname: string, suffix: string): boolean {
  if (suffix === '*') return true;
  return hostname === suffix || hostname.endsWith(`.${suffix}`);
}

function pathMatches(pathname: string, prefix: string | undefined): boolean {
  if (prefix === undefined || prefix === '') return true;
  return pathname === prefix || pathname.startsWith(prefix.endsWith('/') ? prefix : `${prefix}/`) || pathname.startsWith(prefix);
}

/** First matching route wins; otherwise defaultStrategy with optional apifyDefaults for apify strategy. */
export function resolveStrategy(
  config: RoutingConfig,
  urlString: string,
): { strategy: StrategyName; apify?: ApifyRouteConfig } {
  let u: URL;
  try {
    u = new URL(urlString);
  } catch {
    throw new Error(`resolveStrategy: invalid URL: ${urlString}`);
  }
  const hostname = u.hostname.toLowerCase();
  const pathname = u.pathname || '/';

  for (const route of config.routes) {
    const suf = route.match.hostSuffix;
    if (suf !== undefined && !hostMatches(hostname, suf.toLowerCase())) continue;
    if (!pathMatches(pathname, route.match.pathPrefix)) continue;
    let apify: ApifyRouteConfig | undefined;
    if (route.strategy === 'apify') {
      const merged = { ...config.apifyDefaults, ...route.apify };
      if (!merged.actorId) throw new Error('routing: apify route missing actorId and apifyDefaults.actorId');
      apify = merged as ApifyRouteConfig;
    }
    return { strategy: route.strategy, apify };
  }

  const strategy = config.defaultStrategy;
  if (strategy === 'apify') {
    const merged = { ...config.apifyDefaults };
    if (!merged.actorId) throw new Error('routing: default apify strategy requires apifyDefaults.actorId');
    return { strategy, apify: merged as ApifyRouteConfig };
  }
  return { strategy };
}

File: api/src/brain_api/config/load_routing.py
"""Load routing YAML and resolve fetch strategy per URL (mirror cli/src/config/loadRouting.ts)."""

from __future__ import annotations

from typing import Any
from urllib.parse import urlparse

import yaml


def load_routing(yaml_text: str) -> dict[str, Any]:
    raw = yaml.safe_load(yaml_text)
    if not isinstance(raw, dict):
        raise ValueError("routing: root must be a mapping")
    if raw.get("version") != 1:
        raise ValueError("routing: expected version: 1")
    if not raw.get("defaultStrategy"):
        raise ValueError("routing: missing defaultStrategy")
    if not isinstance(raw.get("routes"), list):
        raise ValueError("routing: routes must be an array")
    return raw


def _host_matches(hostname: str, suffix: str) -> bool:
    if suffix == "*":
        return True
    return hostname == suffix or hostname.endswith(f".{suffix}")


def _path_matches(pathname: str, prefix: str | None) -> bool:
    if prefix is None or prefix == "":
        return True
    with_slash = prefix if prefix.endswith("/") else f"{prefix}/"
    return pathname == prefix or pathname.startswith(with_slash) or pathname.startswith(prefix)


def _merge_apify(
    apify_defaults: dict[str, Any] | None,
    route_apify: dict[str, Any] | None,
) -> dict[str, Any]:
    out: dict[str, Any] = {}
    if apify_defaults:
        out.update(apify_defaults)
    if route_apify:
        out.update(route_apify)
    return out


def resolve_strategy(config: dict[str, Any], url_string: str) -> dict[str, Any]:
    try:
        parsed = urlparse(url_string)
    except ValueError as e:
        raise ValueError(f"resolveStrategy: invalid URL: {url_string}") from e
    if not parsed.scheme or not parsed.netloc:
        raise ValueError(f"resolveStrategy: invalid URL: {url_string}")

    hostname = parsed.hostname
    if hostname is None:
        raise ValueError(f"resolveStrategy: invalid URL: {url_string}")
    hostname = hostname.lower()
    pathname = parsed.path or "/"

    routes = config.get("routes") or []
    apify_defaults = config.get("apifyDefaults")

    for route in routes:
        if not isinstance(route, dict):
            continue
        match = route.get("match") or {}
        if not isinstance(match, dict):
            continue
        suf = match.get("hostSuffix")
        if suf is not None:
            if not _host_matches(hostname, str(suf).lower()):
                continue
        path_prefix = match.get("pathPrefix")
        if path_prefix is not None and not isinstance(path_prefix, str):
            path_prefix = str(path_prefix)
        if not _path_matches(pathname, path_prefix):
            continue

        strategy = route.get("strategy")
        if strategy == "apify":
            merged = _merge_apify(
                apify_defaults if isinstance(apify_defaults, dict) else None,
                route.get("apify") if isinstance(route.get("apify"), dict) else None,
            )
            if not merged.get("actorId"):
                raise ValueError(
                    "routing: apify route missing actorId and apifyDefaults.actorId",
                )
            return {"strategy": strategy, "apify": merged}
        return {"strategy": strategy}

    strategy = config["defaultStrategy"]
    if strategy == "apify":
        merged = _merge_apify(
            apify_defaults if isinstance(apify_defaults, dict) else None,
            None,
        )
        if not merged.get("actorId"):
            raise ValueError("routing: default apify strategy requires apifyDefaults.actorId")
        return {"strategy": strategy, "apify": merged}
    return {"strategy": strategy}
